### PR TITLE
fix: webview crash on iframe removal

### LINKED ISSUE #18788: Remove iframe node from <webview> will cause webview crash.
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  *  v5.0.2
* **Operating System:**
  * macOS 10.13.6
* **Last Known Working Electron version:**
  * Haven't tried on other electron versions.

### Expected Behavior
I have an app which runs in electron `<webview>`. I need to render multiple iframes inside this `<webview>` and dynamic add or remove them. 

### Actual Behavior
When I remove one `<iframe>` from the dom tree, the whole `<webview>` crashes. I didn't use JS to remove iframe node, instead, I just delete one `<iframe>` from the dom inspector, and the `<webview>` crashes.

### To Reproduce
This is an example : https://github.com/cloudsere/electron-quick-start. You can run commands below to reproduce.

```sh
$ git clone https://github.com/cloudsere/electron-quick-start
$ npm install
$ npm start
```

### PR DESCRIPTION
**Pull request by alexstrat**
Fixes #18788 #17890

#### Description of Change
Backports this [Chromium fix](https://chromium.googlesource.com/chromium/src/+/7464a6b4e3cf91dd1eb96d922736f0815b94ed9c%5E%21/#F3) as a patch in 5-0-0 branch.

~As a consequence, remove the previous and temporary patch introduced in https://github.com/electron/libchromiumcontent/pull/676~

cc @codebytere @zcbenz @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fix webview crash on iframe removal

### CODE REVIEW

File: patches/common/chromium/portals_fix_crash_when_detaching_oopif_from_portal.patch [hunk 1]
```diff
@@ -0,0 +1,183 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Alexandre=20Lach=C3=A8ze?= <alexandre.lacheze@gmail.com>
+Date: Tue, 25 Jun 2019 17:36:08 +0200
+Subject: Portals: Fix crash when detaching OOPIF from portal.
```
Comment (alexstrat) [inline patches/common/chromium/portals_fix_crash_when_detaching_oopif_from_portal.patch]: I cherry-pick the Chromium commit as it. Not sure it is the expected formatting in the end.

**Comment by codebytere** [comment]
@alexstrat were you able to determine what specifically _about_ this patch breaks the functionality? Ideally i'd love to pare down the patch as much as possible to just the components that are necessary to revert 🤔

**Comment by alexstrat** [comment]
@codebytere my interpretation is that:

Previously, Chromium was unable to detach a subframe of a _(inner) webcontents_ inside a _(outter) webcontens_ (for instance, in our case, a _webview_ in a _BrowserWindow_). That's basically the statement I read in [this comment](https://github.com/electron/electron/issues/14211#issuecomment-418616406) and that led @zcbenz to this [patch](https://github.com/electron/libchromiumcontent/pull/676) (we just don't detach).

The [Chromium patch](https://chromium.googlesource.com/chromium/src/+/7464a6b4e3cf91dd1eb96d922736f0815b94ed9c%5E%21/#F3) seems to fix **exactly** this case: 
> Prior to this change, any detachment of a proxy would cause the inner WebContents to get detached from the outer WebContents.

That's why I would argue to keep the whole patch as it.

That being said, my current knowledge of Chromium architecture did not allow me to:
- understand/find the root cause of the issue, so I'm only an interpretation for now
- understand correctly how the patch in Chromium actually fixes the issue

Therefore, I think the interpretation and opinion of @zcbenz is necessary here.

**Comment by deepak1556** [comment]
Should also try to revert the manual management https://github.com/electron/electron/pull/14487 and see if the issue persists on reload.

**Comment by alexstrat** [comment]
> Should also try to revert the manual management #14487 and see if the issue persists on reload. The new CL seems to address the concerns raised by Cheng's PR, but would be great to get in some tests.

I should rather do that on master, right?

> @alexstrat can you add in tests for the iframe issues that are fixed by the new patch, seems like there are repros that could be turned into a test. Thanks!

Yes, I’ll try that. Though the difficulty is that the webview does not really crash, it just stops to render. So, I still don’t know how to catch that in a test. I’d appreciate ideas!

**Comment by deepak1556** [comment]
> Though the difficulty is that the webview does not really crash, it just stops to render. So, I still don’t know how to catch that in a test. I’d appreciate ideas!

Does that trigger `webContents` `unresponsive` event ? we could listen for that.

**Comment by zcbenz** [comment]
Can you double check this wouldn't cause regression on https://github.com/electron/electron/issues/14211?

**Comment by alexstrat** [comment]
Regarding testing, neither the `'unresponsive'` nor `'crashed'` events are fired when the webview freezes. I tried to hook into the `'paint'` event as well but it does not seem I'm able to catch the paint events for the `webview` and the paint event for the outer webcontents (BrowserWindow) keeps flowing after iframe removal. 😞

**Comment by alexstrat** [comment]
@zcbenz I tried to reproduce #14211 with the instructions provided in description (I waited for ~20 reloads) and could not reproduce it => this PR does cause a regression on #14211

**Comment by zcbenz** [comment]
> I can suggest a PR that reverts #14211 but I think we should 1/ first backport this CL like suggested here so that `5-0-0` matches `6-0-0` and `master`, and 2/ revert #14211 on `master` and backport the revert in `6-0-0` and `5-0-0`.

The plan sounds good to me.

**Comment by deepak1556** [comment]
Can you revert the deletion of `disable_detach_webview_frame.patch` otherwise it doesn't match 6-0-x and master. And as you suggested, a separate PR targeting master can remove this patch along with the manual management, and probably add some tests.

**Comment by alexstrat** [comment]
> Can you revert the deletion of disable_detach_webview_frame.patch otherwise, it doesn't match 6-0-x and master. 

@deepak1556 good catch, yes!

✅done:
- I made sure the CL is the first patch executed
- I took `isable_detach_webview_frame.patch` as it is [just after bumping chromium to 76.0.3783.1]( https://github.com/electron/electron/pull/18116/files#diff-d9e3362ac2c2c5596c5af5895341578e )

**Comment by subbiah95** [comment]
@deepak1556 , I have a similar crash going on in V4.0.0, I am using webview, the crash happened twice,  3 and 5 seconds after start 
the crash points to..
/content/browser/frame_host/render_frame_host_manager.cc
GetProxyToOuterDelegate()->SetChildRWHView(child_rwhv, nullptr);

Is my crash same as this one ?
I am not attaching the dump file as I have changed few codes in electron module(nothing to do with crash) 
electron_crash

**Comment by deepak1556** [comment]
@subbiah95  Can't say for sure without the dump, but I would suggest to try newer versions of electron where the crash might be fixed, V4 is quite old and unsupported release line.

**Pull request merged**